tests/cephadm/test_custom_container.py

```diff
@@ -0,0 +1,272 @@
+
+    for service in services:
+        images = [i for i in out if service in i]
+        config_key = service_config[service]
+        node.exec_command(
+            f"sudo cephadm shell ceph config set mgr mgr/cephadm/{config_key} {node.hostname}:5000/{images[0]}"
+        )
+        out = node.exec_command(
+            f"sudo cephadm shell ceph config get mgr mgr/cephadm/{config_key}"
+        )
+        if node.hostname not in out:
+            raise ClusterCustomContainerError(
+                f"Failed to set custom image mgr/cephadm/{config_key}"
+            )
+
+    # Redeploy the dashboard stack with custom image
+    for service in ["prometheus", "grafana", "alertmanager", "node_exporter"]:
+        node.exec_command(cmd=f"sudo cephadm shell ceph orch redeploy {service}")
+
+        timeout, interval = 30, 5
+        for w in WaitUntil(timeout=timeout, interval=interval):
+            service_info = load(
+                CephAdm(node).ceph.orch.ps(service_name=service, format="json")
+            )
+            image = [c.get("container_image_name") for c in service_info]
+            status = [c.get("status_desc") for c in service_info]
+            if node.hostname not in image and status != "running":
+                log.info(
+                    f"{service} has been re-deployed with custom-image successfully"
+                )
+
+        if w.expired:
+            raise ClusterCustomContainerError(
+                f"Monitor service deployment failed for {service}"
+            )
+
+    return 0
+
+
+def update_cert(ceph):
+    """Check if the certificate is updated on all other nodes in cluster
+    Args:
+        ceph(ceph.ceph.Ceph): CephNode or list of CephNode object
+    """
+
+    out, _ = ceph.exec_command(cmd=f"sudo trust list | grep -i {ceph.hostname}")
+    if ceph.hostname in out:
+        log.info("Certificate added to network node successfully")
+    else:
+        log.error("Certificate failed to add in network node")
+    return 1
+
+
+def start_private_registry(ceph):
+    """Create and start the pricate registry on port 5000
+    Args:
+        ceph(ceph.ceph.Ceph): CephNode or list of CephNode object
+    """
+    ceph.exec_command(
+        cmd=f"sudo podman run --restart=always --name myprivateregistry \
+                    -p 5000:5000 -v {OPT_DATA}:/var/lib/registry:z \
+                    -v {OPT_AUTH}:/auth:z \
+                    -v {OPT_CERTS}:/certs:z \
+                    -e 'REGISTRY_AUTH=htpasswd' \
+                    -e 'REGISTRY_AUTH_HTPASSWD_REALM=Registry Realm' \
+                    -e REGISTRY_AUTH_HTPASSWD_PATH=/auth/htpasswd \
+                    -e 'REGISTRY_HTTP_TLS_CERTIFICATE=/certs/domain.crt' \
+                    -e 'REGISTRY_HTTP_TLS_KEY=/certs/domain.key' \
+                    -e REGISTRY_COMPATIBILITY_SCHEMA1_ENABLED=true \
+                    -d docker.io/meadityaramteke/library"
+    )
+    log.info("Private Registry Started")
+
+
+def create_images_dict(config):
```

Comment:
place the custom methods above the run method.

Reply:
Sure, Pranav! I'll do that.